README.md
# Stack Versions Tracker SVT
  
A small and simple flask and mongo app to keep track of versions of an environment.  
It's a *Proof Of Concept*, it needs to be tested to see what can be improved, secured or else.  
Originally I'm a PHP developper and not an expert in Python, yet I love this language. Also please excuse me, I'm learning ^^.
The concept is to keep it **simple** with only one purpose : have a global view instantly, with easy search.   
  
## Synthesis

Meant to be called with a simple POST call to `/create` by Jenkinsfile, Gitlab-ci or whatever deploy automated process, whenever an app is deployed into an environment.  
This is the minimum required payload :

```
{
    "env": "dev", // can be 'k8s', 'VMs-dev',... and will create a collection in MongoDB if it's not already existing
    "name": "application name", // can't be more clear
    "version": "0.0.1"  // according to specs of https://semver.org, supports build and pre-release label
}
```

**Nota Bene :** The only forbidden characters are:
 - '`$`' and '`.`' : In every 'key' field because of mongo, see [here](https://jira.mongodb.org/browse/SERVER-3229?focusedCommentId=36821&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-36821)
 - '`<`' and '`>`' : In every 'value' field because saved data is going to be displayed in a web page, and I don't want any malicious js or else to be in the db

  
Actually 3 routes that could maybe renamed :  

- '`/create`' method POST [string:env, string:name, string:version] :  
Creates (if not exist, based on name) or updates (if exist, based on name) an application with its name, version, and environment of deployment  
Checks if required params are presents, and also if param version matches a regex.  
Since it\'s mongoDB, no schema and one can add whatever param he likes such as url. Just they won\'t show up
  
- '`/list`' method GET [string:env] :  
Returns list of all applications of an env/collection from DB  

- '`/`' method GET [ ] :  
Shows a template page html with all the entries of one stack  

  
Actually doesn\'t work in standalone, needs to have local mongo server.  

## Docker-compose

I added a docker-compose but not with volume for the application, like I wanted (I could not make the app start).
A simple `docker-compose up` will bring you the application and a mongodb docker.
The volume of mongo works, so the data will be persistent on your disk.

If still you want to dev the app without docker, **keep in mind you will need a mongo DB aside**, but you can run :  
`python app.py`
  
## Config

URL : `http://127.0.0.1:5000`  
MongoDB : `mongodb://127.0.0.1:27017`  
Default database : `stacks`

## Todos & Thoughts

- Make a decent docker-compose that can be sent to prod, examples [here](https://hasura.io/blog/how-to-write-dockerfiles-for-python-web-apps-6d173842ae1d/), [here](http://www.patricksoftwareblog.com/using-docker-for-flask-application-development-not-just-production/), [here](https://netdevops.me/2017/flask-application-in-a-production-ready-container/) and [here](https://flask.palletsprojects.com/en/1.1.x/tutorial/deploy/)
- Manage https as described via [nginx](https://flask.palletsprojects.com/en/1.1.x/deploying/fastcgi/#configuring-nginx) or [here](https://blog.miguelgrinberg.com/post/running-your-flask-application-over-https) 
- Add security, such as [flask-limiter](https://flask-limiter.readthedocs.io/en/stable/)
- Add text field with js and autocompletion for environments selection (fuzzy search mandatory)
- Add datatables to filter results by application name (fuzzy search mandatory)
- Add bootstrap because it's naked
- Here an example of a [complete app well written with flask](https://www.freecodecamp.org/news/structuring-a-flask-restplus-web-service-for-production-builds-c2ec676de563/)
- Add [Flask-restful](https://flask-restful.readthedocs.io/en/latest/) ?
- Add authentication ? UI, API, DB ?
- Manage history by adding each time another version, instead of updating one ?  
 => If yes, create a page to view history of deployment of a certain application ?  
 => Remind that it will lose params like url if not sent, and mongo queries will to be changed

## Tests

 - `test_regex.py` allows to kind of "unit test" the regex for the versions. Probably some tests are missing.

 - In the hypothesis of keeping an history, I tried to insert 260 000 different random entries like the following :

```python
{
	"env":"dev",
	"name":"Name of the application",
	"url":"http://www.whatever.com:1337/random/url/long/enough",
	"version":"5.27.94"
}
```

And the mongo collection weighed about ~ 46 MB.  
  
I choosed 260 000 because it represents :  
365 days - (52 * 2) days of weekend = almost 260 days  

And after it\'s a factor of a 1000 to represent the number of applications, deployments per day and years :  
 - 10 applications deployed 10 times a day over 10 years  
 - 100 applications deployed 10 times a day over 1 year  
 - 1000 applications deployed 1 times a day over 1 year  
 - you choose



## Changelog

[Changelog here](./changelog.md)

changelog.md
# Changelog

1.0.0-RC1
======

 - Dockerized application with use of docker-compose 
 - Changed var name 'env' to 'env_collection' in app.py to avoid confusion
 - Moved application to 'flask_app' folder + added .gitignore


0.10.3
======

 - Added security checks in characters for values inserted in MongoDB

0.10.2
======

 - Added security checks in characters for values inserted in MongoDB

0.10.1
======

 - Added requirements.txt via https://github.com/bndr/pipreqs
 - Prepared env vars for mongo

0.10.1
======

 - Added dropdown to select the environment
 - Added screenshots of poor UI

0.10.0
======

 - Added page to view the result for only one default stack
 - Forced the date on creation and update of an entry


0.9.0 
=====

1st version


flask_app/test_regex.py
import re, pprint
versions = [
"0.0.0",
"0.0.1",
"0.1.2",
# pre-release labels
"9.9.9-aaa",
"9.9.9-AAA",
"9.9.9-Aaa",
"9.9.9-Aaa123",
# build labels
"9.9.9+aaa",
"9.9.9+AAA",
"9.9.9+Aaa",
"20.19.4+Abcdef145",
# pre-release AND build labels
"0.9.4-lowercase+build-label",
"0.9.4-l0w3rc4s3-and-numbers+build-label",
"0.9.4-UPP3RCASE-AND-NUMBERS+buidlowercase",
"0.9.4-UPPERCASE+buidlowercase",
"0.9.4-UPPERCASEPRERELEASE+BUILDUPPERCASE",
"0.9.4-TEST1235+Akindof23.Bad.Test",
"0.9.4-TEST1235+Another123*bAd*Test",
"0.9.4-TEST1235+Another123 baD Test",
# bad versions
" 0.9.4",
"0. 9.4",
"0.9 .4",
"0.9.4 ",
"µ/0.9.4",
"+0.9.4",
"0.9+.4",
"0.9.4+",
"-0.9.4",
"0.9-.4",
"0.9.4-",
"a.9.4",
"0.b.4",
"0.9.c",
"0.9.4//..",
"a.b.c",
"a.9.4",
"(..9.4",
"...4",
]

# You can go test your version against the pattern with tools like https://regex101.com/
for ver in versions :
	if re.search("^[0-9]+\.[0-9]+\.[0-9]+[+0-9A-Za-z-]*$", ver):
	    # pprint.pprint("Version '"+ver+"' / OK")
	    pprint.pprint("Version '{}' :  OK".format(ver))
	else:
	    pprint.pprint("Version '"+ver+"' <------ Bad version")



flask_app/test_requests.py
import requests

url = 'http://127.0.0.1:5000/create'
headers = {'Content-Type': "application/json", 'Accept': "application/json"}

print('=============')
print('Good requests')
print('=============')
good = [
    '{"env": "dev", "name":"Test app","url":"https://url.of.test.com","version":"1.0.0-RC1"}',
    '{"env": "dev", "name":"Test app","url":"https://url.of.test.com","version":"1.0.1-beta+12sa5s4d4ze"}',
    '{"env": "dev", "name":"Second app","url":"https://other.url.com:1337","version":"1.0.2"}'
]

for payload in good:
    res = requests.post(url, data=payload, headers=headers,)
    if res.status_code == 200:
        print("Expected response 200, recieved "+str(res.status_code)+" : OK")
    else:
        print("Unexpected response, should be 200, is " + str(res.status_code))
        print("Payload : \n" + payload)



print('============')
print('Bad requests')
print('============')
bad = [
  '{"e$nv": "dev", "name":"Test app","url":"https://url.of.test.com","version" : "1.0.3"}',
  '{"env.": "dev", "name":"Test app","url":"https://url.of.test.com","version" : "1.0.3"}',
  '{"env": "<dev>", "name":"Test app","url":"https://url.of.test.com","version" : "1.0.3"}',
  '{"env": "dev", "name":"Test app<","url":"https://url.of.test.com","version" : "1.0.3"}',
  '{"env": "dev", "name":"Test app","url":"https://url.of.test.com>","version" : "1.0.3"}',
  '{"env": "dev", "name":"Test app","url":"https://url.of.test.com","version" : ">1.0.3"}',
  '{"env": "dev", "name":"Test app","url":"https://url.of.test.com","version" : "1.0.3 "}',
  '{"env": "dev", "name":"Test app","url":"https://url.of.test.com","version" : "-1.0.3"}',
  '{"env": "dev", "name":"Test app","url":"https://url.of.test.com","version" : "1.0.3:"}'
]

for payload in bad:
    res = requests.post(url, data=payload, headers=headers,)
    if res.status_code == 400:
        print("Expected response 400, recieved "+str(res.status_code)+" : OK")
        print(res.content)
    else:
        print("Unexpected response, should be 400, is " + str(res.status_code))
        print("Payload : \n" + payload)
        print(res.content)


flask_app/app.py
from flask import Response, Flask, render_template, request,redirect,url_for
from bson.json_util import dumps
from pymongo import MongoClient
from datetime import datetime
from utils.vars import *
from config import *
import re, pprint, json


app = Flask(__name__)

client = MongoClient(MONGO_URI)
db = client.stacks

@app.route("/", methods=['GET', 'POST'])
@app.route("/index", methods=['GET', 'POST'])
def index():
    if request.form:
        env_collection = request.form['environment']
    else: # default value for collection, needs to be rethinked with empty DB
        available_collections = db.list_collection_names()
        if not available_collections:
            return render_template('index.html')
        env_collection = available_collections[0]
    response = findAllEntriesForEnv(env_collection)
    # Route for template rendering
    return render_template(
        'index.html',
        data=response,
        envs=db.list_collection_names(),
        selected_env=env_collection
    )

@app.route("/list/<env_collection>")
def list (env_collection): # Check characters for security here ?
    response = findAllEntriesForEnv(env_collection)
    return Response(response, status=200, mimetype='application/json')


@app.route("/create", methods=['POST'])
def createVersion ():
    # Payload presence check
    if not request.data:
        return Response(expected_create_body, status=400, mimetype='application/json')
    json_payload = json.loads(request.data)

    # Securing payload
    if len(json_payload) > 10: # TODO : Check what value to put here
        return Response("Too many elements in payload", status=400, mimetype='application/json')
    for key in json_payload:
        if key.find("$") != -1 or key.find(".") != -1: # Mongo-related security, see README.md
            return Response('"$" or "." character not allowed in "'+key+'":"'+json_payload[key]+'"', status=400, mimetype='application/json')
        if json_payload[key].find("<") != -1 or json_payload[key].find(">") != -1 : # To avoid arbitrary html sent to app when showing in template
            return Response('"<" and ">" characters not allowed in "'+key+'":"'+json_payload[key]+'"', status=400, mimetype='application/json')

    # Validating presence of params
    for param in required_params:
        if not param in json_payload: # checks if param exists
            return Response("Missing required param : "+param+"\n"+expected_create_body, status=400, mimetype='application/json')
        if not json_payload[param]: # checks if param is empty
            return Response("Param '"+param+"' can't be empty \n"+expected_create_body, status=400, mimetype='application/json')

    # Assigning mandatory params to values
    env_collection = json_payload['env']
    name = json_payload['name']
    version = json_payload['version']

    # Validating format of version according to https://semver.org
    if not re.search(VERSION_REGEX_PATTERN, version):
        print("Version does not match regex")
        return Response("Version '"+version+"' must match regex : {}".format(VERSION_REGEX_PATTERN), status=400, mimetype='application/json')
    
    # Version matches regex, so we can save the payload in database
    else:
        # Selecting the collection from request
        stack = db[env_collection]

        # Find any application with the same name in the collection
        counted_results = stack.count_documents({"name": {"$regex" : "^" + name + "$"}})

        if counted_results > 1 :
            return Response("More than one application with the name : " + name, status=409, mimetype='application/json')
            
        elif counted_results < 1 : # no entry, so we create
            json_payload['date'] = datetime.now() # add time of creation
            stack.insert_one(json_payload)
            return Response("Creation successful", status=200, mimetype='application/json')

        else: # exactly one entry, so we update
            json_payload['date'] = datetime.now()  # update time of deployement
            stack.update_one({"name": name}, { "$set" : json_payload})
            return Response("Update successful", status=200, mimetype='application/json')
    
    return Response("Unexpected error", status=418, mimetype='application/json')


def findAllEntriesForEnv(env_collection):
    stack = db[env_collection]
    return dumps(stack.find())

if __name__ == "__main__":
    app.run(use_reloader=True,host='0.0.0.0',port=5000)